plugins/experimental/carp/CarpHashAlgorithm.cc

```diff
@@ -0,0 +1,396 @@
+  
+  HashNode* node = findStatusByNameAndPort(name,port);
+  if(node) {
+    node->setStatus(status,time,latencyMs);
+  } else {
+    TSError("Carp internal error setStatus host %s not found",name.c_str());
+  }
+}
+
+void
+HashAlgorithm::setStatus(HashNode * node, bool status, time_t time, uint64_t latencyMs)
+{
+  TSDebug(DEBUG_TAG_INIT, "HashAlgorithm::setStatus name=%s status=%d", node->name.c_str(), status);
+
+//  HashNode* node = findStatusByNameAndPort(name,port);
+  if(node) {
+    node->setStatus(status,time,latencyMs);
+  } else {
+    TSError("Carp internal error setStatus host %s not found",node->name.c_str());
+  }
+}
+
+/*****************************************************************/
+HashAlgorithm::~HashAlgorithm()
+{
+  for (size_t ptr = 0; ptr < _hostList.size(); ptr++) {
+    delete _hostList[ptr];
+  }
+}
+
+/*****************************************************************/
+void
+HashAlgorithm::dump(string& s)
+{
+  _config->dump(s);
+  
+  stringstream ss;
+  ss << "Hash Algo stats:" << endl;
+  s += ss.str();
+  
+  for (size_t ptr = 0; ptr < _hostList.size(); ptr++) {
+    ss.str("");
+    ss.clear();
+    ss << ptr << "-";
+    s += ss.str();
+    _hostList[ptr]->dump(s);
+  }
+}
+
+/*****************************************************************/
+void
+CarpHashAlgorithm::algoInit()
+{
+  // calculate hash for each host
+  for (size_t ptr = 0; ptr < _hostList.size(); ptr++) {
+    std::string cname;
+    char buf[64];
+    memset(buf, 0, 64);
+    snprintf(buf, 64, "%s:%d", _hostList[ptr]->name.c_str(), _hostList[ptr]->listenPort);
+    cname = buf;
+
+    _hostList[ptr]->hash = _calculateHash(cname);
+  }
+
+  // calculate loadmultiplier
+  _calculateLoadMultiplier();
+  
+   for (size_t ptr = 0; ptr < _hostList.size(); ptr++) {
+     TSDebug(DEBUG_TAG_INIT, "algoInit host=%s port=%d hash=0x%x weight=%g loadFac=%g loadMult=%g isSelf=%d status=%d", 
+             _hostList[ptr]->name.c_str(),
+             _hostList[ptr]->listenPort,
+             _hostList[ptr]->hash,
+             _hostList[ptr]->weight,
+             _hostList[ptr]->loadFactor,
+             _hostList[ptr]->loadMultiplier,
+             _hostList[ptr]->isSelf,
+             _hostList[ptr]->getStatus());
+   }
+}
+
+/*****************************************************************/
+std::vector<HashNode*>
+CarpHashAlgorithm::getRemapProxyList(const std::string& url)
+{
+  std::vector<HashNode *> hn=_selectReplicateNodes(url);
+  return hn;
+}
+
+/*****************************************************************/
+HashNode*
+CarpHashAlgorithm::getRemapProxy(const std::string& url)
+{
+  HashNode *hn=_selectNode(url);
+  if(hn) {
+    hn->hit();
+  }
+  return hn;
+}
+
+
+/*****************************************************************/
+unsigned int
+CarpHashAlgorithm::_calculateHash(const std::string& hostname)
+{
+  // check input data
+  if (hostname.empty())
+    return 0;
+
+  unsigned int hash = 0;
```

Comment:
Does the hash output have a well defined size? That is, is it specifically a 32 bit or 64 bit (for example) hash?